src/twisted/protocols/_smb/_base.py

```diff
@@ -0,0 +1,396 @@
+     - C{ERROR} throw a L{SMBError}
+     - C{OFFSET} return offset into data
+     where remainder begins as second item of tuple
+     - C{DATA} return remaining data as second item of tuple
+    @type remainder: L{int}
+
+    @param offset: offset into data to begin from
+    @type offset: L{int}
+
+    @return: an instance of C{cls}, or a 2-tuple, first item the former,
+             second as determined by C{remainder}
+    """
+    strct = _get_struct(cls)
+    if remainder == ERROR and strct.size + offset < len(data):
+        raise SMBError("unexpected remaining data")
+    ret = strct.unpack_from(data, offset=offset)
+    fields = smb_fields(cls)
+    assert len(fields) == len(ret)
+    kwargs = {}
+    for i in range(len(ret)):
+        val = ret[i]
+        if fields[i].type is uuid_mod.UUID:
+            val = uuid_mod.UUID(bytes_le=val)
+        kwargs[fields[i].name] = val
+    obj = cls(**kwargs)
+    if remainder <= ERROR:
+        return obj
+    elif remainder == OFFSET:
+        return (obj, offset + strct.size)
+    else:
+        return (obj, data[offset + strct.size :])
+
+
+def smb_fields(cls: Any) -> List[Any]:
+    return [i for i in attr.fields(cls) if SMB_METADATA in i.metadata]
+
+
+def _get_struct(cls: Any) -> struct.Struct:
+    try:
+        # we use classes to hold cache of Structs as precompiling is more
+        # efficient
+        strct = cast(struct.Struct, cls._struct)
+    except AttributeError:
+        strct = struct.Struct(
+            "<" + "".join(i.metadata[SMB_METADATA] for i in smb_fields(cls))
+        )
+        cls._struct = strct
+    return strct
+
+
+def calcsize(cls: Any) -> int:
+    """
+    return the size of a structure.
+    @param cls: C{attr} decorated class
+    @rtype: L{int}
+    """
+    strct = _get_struct(cls)
+    return strct.size
+
+
+_leint = struct.Struct("<I")
+
+
+def int32key(d: Dict[int, Any], val: Any) -> int:
+    """
+    generate a new random key for a dictionary
+    @param d: dictionary with 32-bit int keys
+    @type d: L{dict}
+    @param val: new dictionary value
+    @rtype: L{int}
+    """
+    assert len(d) < 0xC0000000  # otherwiae dict so big hard to find keys
+    n = 0
+    while n == 0 or n in d:
+        [n] = _leint.unpack(secureRandom(_leint.size, True))
+    d[n] = val
+    return n
+
+
+@attr.s
+class SMBPacket:
+    """
+    A SMB packet as it moves through the system.
+
+    @ivar data: raw data of the packet, both reception and
+                transmission.
+    @type data: L{bytes}
+
+    @ivar hdr: the parsed header
+    @ivar body: the parsed body
+    """
+
+    data: bytes = attr.ib()
+    _proto: "SMBPacketReceiver" = attr.ib()
+    hdr: Any = attr.ib(default=None)
+    body: Any = attr.ib(default=None)
+
+    @property
+    def ctx(self) -> Dict[str, Any]:
+        """
+        the connection context: objects that need to persist scross
+        packets
+        @rtype: L{dict}
+        """
+        return self._proto.ctx
```

Comment:
Clients of this class appear to have very strong expectations on the types of the values that get added to this dictionary. Should this be a user-supplied class, or collection of classes? `Dict[str, Any]` feels like way too broad of a type, particularly given that this seems like something that will be exposed to applications via `packetReceived`.